vnext/Microsoft.ReactNative/Views/TextViewManager.cpp

```diff
@@ -101,92 +94,36 @@ class TextShadowNode final : public ShadowNodeBase {
     const auto textBlock = this->GetView().as<xaml::Controls::TextBlock>();
     textBlock.TextHighlighters().Clear();
 
-    auto nestedIndex = 0;
-    if (m_hasDescendantBackgroundColor) {
-      const auto highlighterCount = textBlock.TextHighlighters().Size();
-      if (const auto uiManager = GetNativeUIManager(GetViewManager()->GetReactContext()).lock()) {
-        for (auto childTag : m_children) {
-          if (const auto childNode = uiManager->getHost()->FindShadowNodeForTag(childTag)) {
-            nestedIndex = AddNestedTextHighlighter(
-                m_backgroundColor, m_foregroundColor, static_cast<ShadowNodeBase *>(childNode), nestedIndex);
-          }
+    // Since TextShadowNode is not public, we lift some of the recursive
+    // algorithm into the shadow node implementation to detect when no
+    // descendants have background colors and we can skip recursion.
+    if (m_hasDescendantTextHighlighter) {
+      const auto highlighters = GetNestedTextHighlighters(this, m_foregroundColor, m_backgroundColor);
+      if (highlighters.size() == 0) {
+        m_hasDescendantTextHighlighter = false;
+      } else {
+        // We must add the highlighters in reverse order, as highlighters
```

Comment:
nit: can a stack be used instead of a vector for this?

Reply:
Probably. Is it more efficient to do so?